src/profitability.py

```diff
@@ -72,11 +82,118 @@ def profitability(self, fiat: fiat, hashrate: int, power_consumption: int, elect
+        results.append(self._request_headers_range(start_height, end_height))
+        result = pd.concat(results, axis=0)
+        return result
+    
+    def historical_diff(self, height:int=None, timestamp:datetime=None, batch_size:int=1000) -> int:
+        if height and timestamp:
+            warnings.warn("Both height and timestamp present: ignoring timestamp", UserWarning)
+        
+        path = f"{DIR_TMP}/diff_{self.coin.name}.pkl"
+        try: # If we have previous saved data, merge with the new data
+            diff = pd.read_pickle(path)
+        except FileNotFoundError:
+            print(f"{path} does not exist. Creating...")
+            # diff = self._request_headers_batcher(0, self.height, batch_size=batch_size)
+            diff = self._request_headers_batcher(0, height, batch_size=batch_size)
+            diff.to_pickle(path)
+            print(diff)
+            pass
+        
+        last_known_height = diff.index[-1]
+        last_known_timestamp = datetime.fromtimestamp(diff["timestamp"].iloc[-1])
+        print(last_known_timestamp)
+        
+        if height:
+            if height > self.height:
+                raise ValueError("Requested height is in the future")
+            if height < 0:
+                raise ValueError("Cannot have a negative height")
+            if height > last_known_height:  # Need to update
+                # diff_new = self._request_headers_batcher(last_known_block + 1, self.height, batch_size=batch_size)
+                diff_new = self._request_headers_batcher(last_known_height + 1, height, batch_size=batch_size)
+                diff = pd.concat([diff, diff_new], axis=0)
+                diff.to_pickle(path)
+            return diff.at[height, "difficulty"]
+        elif timestamp:
+            raise NotImplementedError("WIP")
+            if timestamp > datetime.now():
+                raise ValueError("Requested timestamp is in the future")
+            if timestamp < datetime.fromtimestamp(0):
+                raise ValueError("Cannot have a negative timestamp")
+            if timestamp > last_known_timestamp:  # Need to update
+                target_height = min(self.height, last_known_height + (timestamp - last_known_timestamp) / 120)
+                # diff_new = self._request_headers_batcher(last_known_height + 1, self.height, batch_size=batch_size)
+                diff_new = self._request_headers_batcher(last_known_height + 1, target_height, batch_size=batch_size)
+            # search timestamps and find nearest-previous height index
+            height = diff.loc[diff.index.get_loc(timestamp, method="nearest")]
+            return diff.at[height, "difficulty"]
+        else:  # not height and not timestamp:
+            raise TypeError("Need a height or a timestamp")
+
 
 def test():
+    import time
     a = POW_Coin(coin.XMR)
     b = a.profitability(fiat.USD, 20000, 200, 0.1)
-    print(b)   
+    print(b)
```

Comment:
Please use asserts for testing.

This assuming that there was no connection problem.